rpc: Filter inputs by type during CoinSelection

aureleoules opened the pull request:
This PR adds a filter to the CoinControl to select only specific utxos by type. It allows the `fundrawtransaction` and `walletcreatefundedpsbt` rpc calls to filter inputs by type.

Closes #25181.

aureleoules committed:
rpc: Filter inputs by type during CoinSelection

theStack commented:
Concept ACK

S3RK commented:
@aureleoules can we achieve the same result by using descriptors wallet with only segwit descriptors?

brunoerg reviewed (commented):
Concept ACK

I think you should squash the commits since the test won't work without the implementation.

promag commented:
Concept ACK.

@brunoerg I think it's fine as it. Should squash if the 1st commit requires the 2nd to pass CI.

aureleoules commented:
> @aureleoules can we achieve the same result by using descriptors wallet with only segwit descriptors?

I am not very familiar with wallet descriptors.
Do you mean using a wallet that does not have legacy descriptors imported?
I think @achow101 answed that yes (https://github.com/bitcoin/bitcoin/issues/25181#issuecomment-1134923482).

But if I understand correctly, this feature might still be useful in cases where you don't control the wallet.

luke-jr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -787,7 +792,8 @@ RPCHelpMan fundrawtransaction()
                                 },
-                             },
+                            },
+                            {"inputs_witness_only", RPCArg::Type::BOOL, RPCArg::Default{false}, "Whether to only use inputs with witness data for transaction."},
```
The name here seems confusing to me. Maybe `"segwit_inputs_only"`?

aureleoules replied:
Indeed, I changed it.

Xekyo commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -177,6 +178,38 @@ def test_simple(self):
+        inputs  = [ ]
+
+        # make sure legacy inputs are not accepted in witness only mode if no witness inputs are found
+        self.generatetoaddress(self.nodes[2], 2, self.nodes[2].getnewaddress(address_type='legacy'))
```
This may actually be testing something else than intended: given that the coinbase output is not 100 blocks old yet, I suspect the "Insufficient funds" would be foremost caused by the inputs being immature.

This test could be improved by creating a scenario in which a transaction is first successfully funded by using legacy inputs, but then toggling the `'segwit_inputs_only': true` causes a repeated attempt to fund the transaction to fail.

murchandamus reviewed (commented):
I noticed that this PR changes similar code parts as #24584. Perhaps, the changes to `availableCoins` in #24584 could be useful in implementing the idea a bit more elegantly. It may be good to coordinate this PR with the author of #24584, perhaps #25183 should build on #24584 or vice versa.

Newman1515 reviewed (commented):
Submit the approve

ishaanam commented:
Would it be useful to add the `"segwit_inputs_only"` option to `walletcreatefundedpsbt` as well? They both call `FundTransaction` so I think you would just have to include it in the `walletcreatefundedpsbt` options.

aureleoules commented:
I've rebased my PR on top of #24584 as @Xekyo suggested and addressed the comments.
Thanks for the reviews!

S3RK commented:
> But if I understand correctly, this feature might still be useful in cases where you don't control the wallet.

@aureleoules Do you have a concrete use-case in mind? I'm not sure, but it seems #25181 supposes that we do control the wallet.

furszy commented:
Instead of adding the `segwit_inputs_only` filter only, what if we add a general input type filter? Something like `filter_inputs_by=<OutputType>`

So we aren't forced to add a new `type*_inputs_only` flag in every RPC command every time that a new inputs type filter is required.

aureleoules commented:
> Do you have a concrete use-case in mind? I'm not sure, but it seems https://github.com/bitcoin/bitcoin/issues/25181 supposes that we do control the wallet.

I believe if your wallet already has mixed inputs and you want to filter these, this feature may be useful.

fanquake commented on src/wallet/coincontrol.h:
```diff
@@ -12,6 +12,7 @@
 #include <script/keyorigin.h>
 #include <script/signingprovider.h>
 #include <script/standard.h>
+#include <unordered_set>
```
This should go with the other std lib includes.

aureleoules replied:
Fixed, thanks.

aureleoules commented:
> Instead of adding the `segwit_inputs_only` filter only, what if we add a general input type filter? Something like `filter_inputs_by=<OutputType>`
> 
> So we aren't forced to add a new `type*_inputs_only` flag in every RPC command every time that a new inputs type filter is required.

Thanks for the feedback, I've added it.

josibake committed:
wallet: add filter to AvailableCoins

add m_filter_inputs to coinControl, which then
allows us to filter by OutputType in AvailableCoins

josibake commented:
Concept ACK

This is coming along nicely! I do have a few suggestions:

I think the correct place to do the filtering is in `AvailableCoins`. This would make it more useful outside of just funding a transaction. I recently opened #25734 which should make this _much_ easier. All we would need is your update to `coincontrol.h` and a small change to `AvailableCoins` and then any RPC can filter by just adding filters to `m_filter_inputs` (like you've done with `fundrawtransaction`.

I rebased this off #25734 , added my suggested changes, and verified that the functional tests still ran. Feel free to cherry pick:

* [josibake@e4369e6](https://github.com/bitcoin/bitcoin/commit/e4369e604ad278741a0d50136e3f8dbca02e390c) (add filter to AvailableCoins)
* [josibake@2609ab1](https://github.com/bitcoin/bitcoin/commit/2609ab114e7e8b5c8ef9d85796d9851d0c3da612) (your commit rebased on my changes)

Also, small nit (feel free to ignore): I'd suggest breaking up your commits to make it easier to review.. something like 1) make changes to AvailableCoins 2) make changes to the RPC 3) make changes to the functional test. Makes it much easier from a review perspective, imo

josibake commented on src/wallet/rpc/spend.cpp:
```diff
@@ -486,6 +486,27 @@ static std::vector<RPCArg> FundTxDoc(bool solving_data = true)
+    return *outputType;
+}
+
+static inline std::vector<std::string> parseStringList(std::stringstream& ss)
+{
+    std::vector<std::string> result;
+    std::string item;
+    while (ss.good()) {
+        getline(ss, item, ',');
+        result.push_back(item);
+    }
+    return result;
+}
```
this function should also handle lists like `bech32, bech32m`, as most users will probably pass them with spaces. i tested that this currently fails by passing `bech32, bech32m` in the functional test.

aureleoules replied:
Thanks, I fixed it

josibake commented on src/wallet/coinselection.h:
```diff
@@ -145,6 +147,8 @@ struct CoinSelectionParams {
      * reuse. Dust outputs are not eligible to be added to output groups and thus not considered. */
     bool m_avoid_partial_spends = false;
+    /** Filter transaction inputs by type */
+    std::unordered_set<OutputType> m_filter_inputs;
```
if you do the filtering in `AvailableCoins`, we don't need to add this to `coinselection.h`

josibake commented on src/wallet/spend.cpp:
```diff
@@ -776,6 +801,7 @@ static BResult<CreatedTransactionResult> CreateTransactionInternal(
 
     CoinSelectionParams coin_selection_params{rng_fast}; // Parameters for coin selection, init with dummy
     coin_selection_params.m_avoid_partial_spends = coin_control.m_avoid_partial_spends;
+    coin_selection_params.m_filter_inputs = coin_control.m_filter_inputs;
```
same as above, this isn't needed if we filter in `AvailableCoins`

josibake commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -181,6 +182,41 @@ def test_simple(self):
+
+        return True
+
+    def test_witness_only(self):
```
it would be great to test the case where we filter inputs to a particular type (or multiple types) and then verify that the inputs in the funded tx only contain those types. you can look at `test/functioanal/wallet_avoid_mixing_output_types.py` for an example of checking the inputs to a transaction and verifying they are all of a certain `OutputType`

aureleoules replied:
Thanks, I've added these tests.
I included some helpers from `wallet_avoid_mixing_output_types` to the test framework to reuse them.

aureleoules commented:
Thank you @josibake for the review and the commits! I added your commits and fixed the list parsing.
I will address the test a bit later.

aureleoules commented:
I have rebased the PR on top of @josibake's PR #25734.
The RPC `fundrawtransaction` and `walletcreatefundedpsbt` now take an option to filter inputs during the coin selection.
I've added a test case.

furszy commented on src/wallet/coincontrol.h:
```diff
@@ -38,6 +39,8 @@ class CCoinControl
     //! while requires all selected inputs be used
     bool m_allow_other_inputs = false;
+    //! Filter inputs by type
+    std::unordered_set<OutputType> m_filter_inputs;
```
Small idea, what about using an optional bit-field here?
So you later can skip coins with a fast:
```c++
if (m_filter_inputs && !(*m_filter_inputs & coin_type)) continue;
```

aureleoules replied:
great idea, I pushed it
`git diff 0a9d345 89ea8fc`

aureleoules commented on src/wallet/rpc/spend.cpp:
```diff
@@ -530,13 +552,24 @@ void FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& fee_out,
+            const auto& outputTypes = parseStringList(ss);
+            coinControl.m_filter_inputs = 0;
+            for (size_t i = 0; i < outputTypes.size(); i++) {
+                *coinControl.m_filter_inputs |= static_cast<unsigned int>(parseOutputType(outputTypes[i]));
```
thanks, i pushed it

S3RK commented:
> > Do you have a concrete use-case in mind? I'm not sure, but it seems #25181 supposes that we do control the wallet.
> 
> I believe if your wallet already has mixed inputs and you want to filter these, this feature may be useful.

I was thinking about this proposal and I'm still not convinced there is a clear understanding how this would be actually used in the real world.

The declared benefit is non-malleability for lightning wallets (#25181). For new lightning wallets this is a non-issue as they shouldn't have non-segwit descriptors in the first place. IIUC the proposed solution for existing wallets is to filter malleable inputs when creating channel opening txs in `fundrawtransaction` and `walletcreatefundedpsbt`. This approach renders a portion of wallet funds unusable and have a number of problems:
1) users could be confused about funding errors despite having enough balance
2) it doesn't provide a solution how to make all the funds usable
3) what about other RPCs? what if non-malleable inputs are added by `walletprocesspsbt` or when bumping psbt?
     
It seems what you really want is to "unmix" the wallet, i.e. sweep all malleable inputs as a one-time operation. 
Today, I think, there are two main approaches users can take to achieve this:
1) (preferred, descriptors only) create a new wallet and import only segwit descriptors from the old wallet
2) create a new lighting wallet and fund it with desired amount to one or more addresses 

While 2nd approach has some downsides, maybe it's just good enough. I don't know whether there are enough users affected and whether these downsides outweigh the added complexity.
